src/simulation/locomotionAlgorithms/dijkstra.py

```diff
@@ -0,0 +1,64 @@
+from typing import List
+from src.enums.tileStatus import TileStatus
+from src.simulation.destination.pedestrianHeapMap import PedestrianHeapMap
+from src.exceptions.simulationException import SimulationException
+from src.exceptions.simulationErrorCodes import SimulationErrorCodes
+from src.simulation.grid import Grid
+from src.simulation.tile import Tile
+
+import numpy as np
+
+class Dijkstra():
+
+    def __init__(self, grid: Grid, destinationTiles: List[tuple]) -> None:
+        #Dont use the constructor directly
+        self.grid = grid
+        self.destinationTiles = destinationTiles
+        self.stack: List[tuple] = []
+        self.pedestrianHeatMap: PedestrianHeapMap = PedestrianHeapMap(grid.length, grid.width)
+
+    def calculatePedestrianHeapMap(self) -> PedestrianHeapMap:
+        self._setStartNodes()
+        while len(self.stack) > 0:
+            currentNode = self._getNextNode()
+            self._evaluateNode(currentNode)
+        return self.pedestrianHeatMap
+
+    def _setStartNodes(self):
+        for x, y in self.destinationTiles:
+            self.pedestrianHeatMap.updateValue(x, y, 0)
+            self.stack.append((x, y, 0))
+
+    def _getNextNode(self) -> tuple:
+        returnTuple = min(self.stack, key=lambda x: x[2])
+        self.stack.remove(returnTuple)
+        return returnTuple
+    
+    def _evaluateNode(self, currentNode) -> None:
+        for newX in range(currentNode[0] - 1, currentNode[0] + 2):
+            for newY in range(currentNode[1] - 1, currentNode[1] + 2):
+                if self.grid.contains(newX, newY) and self.grid.getTileStatus(newX, newY) != TileStatus.BLOCKED:
+                    distance = 1 if self._isDiagonalStep(currentNode[0], currentNode[1], newX, newY) else np.sqrt(2)
+                    newValue = currentNode[2] + distance
+                    if (self.pedestrianHeatMap.getValue(newX, newY) > newValue):
+                        self.pedestrianHeatMap.updateValue(newX, newY, newValue)
+                        self.stack.append((newX, newY, newValue))
+    
+    def _isDiagonalStep(self, x:int, y:int, xNew:int, yNew: int) -> bool:
```

Comment:
sure this method does what it says?

Reply:
Yes we are. The Step is not diagonal, one of the indices (x or y) is the same number as the new indices(xNew, yNew) therefore either x -xNew or y - yNew is zero.

Reply:
The method checks if either the substraction of the x coordintes are zero or the substraction of y coordintes. If one of each is true, it is NOT diagonal. Imho, the method should be named differently or should test that neither the  x substraction nor y substraction are allowed to be zero.

Reply:
I'm sorry. Well spotted. Thank you.